### PR TITLE
Adding generic SAML support

### PR DESCRIPTION
**Pull request by pieterlukasse**
# What? Why?
This PR introduces a generic SAML solution in the form of a new `SAMLUserDetailsServiceImpl.java` (as an alternative to the current MSKCC custom `MSKCCPortalUserDetailsService.java`) and a set of new properties to support different SAML IDP - SP integration scenarios. 

# Changes proposed in this pull request:
- new `SAMLUserDetailsServiceImpl.java` (as an alternative to the current MSKCC custom `MSKCCPortalUserDetailsService.java`) 
- new set of `saml` properties for `portal.properties.EXAMPLE` 
- documentation of the purpose and use of the new properties

# Notify reviewers
@n1zea144 @inodb

### CODE REVIEW

**Comment by pieterlukasse** [comment]
@n1zea144 here is the SAML solution we discussed some time ago.

File: docs/Authenticating-Users-via-SAML.md [hunk 1]
```diff
@@ -18,15 +18,18 @@ In its simplest terms, SAML boils down to four terms:
...
 
 # Setting up an Identity Provider
 
-As noted above, we provide details on how to preform SAML authentication via a commercial company:  [OneLogin](https://www.onelogin.com/).  OneLogin provides a free tier for testing out SAML authentication, and is one of the easier options to get a complete SAML workflow set-up.  Once you have OneLogin working, you should then have enough information to transition to your final authentication service.  As you follow the steps below, the following link may be helpful: [How to Use the OneLogin SAML Test Connector](https://support.onelogin.com/hc/en-us/articles/202673944-How-to-Use-the-OneLogin-SAML-Test-Connector).
+As noted above, we provide details on how to preform SAML authentication via a commercial company:  [OneLogin](https://www.onelogin.com/).
```
Comment (pieterlukasse) [inline docs/Authenticating-Users-via-SAML.md]: done

File: docs/Authenticating-Users-via-SAML.md [hunk 2]
```diff
@@ -37,16 +40,30 @@ To get started:
 * [Login to OneLogin.com](https://app.onelogin.com/login).
 * Under Apps, Select Add Apps.
 * Search for SAML.
-* Select the option labeled: OneLogin SAML Test (IdP).
+* Select the option labeled: OneLogin SAML Test (IdP w/attr).
+
+
+* Under the Configuration Tab for OneLogin SAML Test (IdP w/attr), paste the following fields (this is assuming you are testing everything via localhost).
```
Comment (pieterlukasse) [inline docs/Authenticating-Users-via-SAML.md]: done

File: docs/Authenticating-Users-via-SAML.md [hunk 3]
```diff
@@ -37,16 +40,30 @@ To get started:
...
 
-    * ACS (Consumer) URL Validator*:  ^http:\/\/localhost\:8080\/cbioportal\/$
-    * ACS (Consumer) URL*:  http://localhost:8080/cbioportal/
+* Add at least the parameters: 
+    * Email (Attribute)
+    * Email (SAML NameID)
+
+![](images/previews/onelogin-config-parameters.png)
+
+* Configure these **email** parameters in the Users menu:
```
Comment (pieterlukasse) [inline docs/Authenticating-Users-via-SAML.md]: Thanks. Added some more steps and 2 screenshots to clarify this part.

File: docs/Authenticating-Users-via-SAML.md [hunk 4]
```diff
@@ -92,52 +109,68 @@ Within portal.properties, make sure that:
...
+    saml.logout.url=/
 
-Please note that you will have to modify all the above to match your own settings.
+Please note that you will have to modify all the above to match your own settings. `saml.idp.comm.binding.type` can be left empty if `saml.idp.comm.binding.settings=defaultBinding`. The `saml.logout.*` settings above reflect the settings of an IDP that supports Single Logout (hopefully the default in most cases - more details in section below).
 
-## Authorizing Users
+### Custom scenarios
 
-Next, please read the Wiki page on [User Authorization](User-Authorization.md), and add user rights for a single user.
+:information_source: some settings may need to be adjusted to non-default values, depending on your IDP. For example, of your
```
Comment (pieterlukasse) [inline docs/Authenticating-Users-via-SAML.md]: fixed

File: docs/Authenticating-Users-via-SAML.md [hunk 5]
```diff
@@ -92,52 +109,68 @@ Within portal.properties, make sure that:
...
 
-This step requires that you make a modest amount of code changes.  You can either choose to modify:  [MSKCCPortalUserDetailsService.java](https://github.com/cBioPortal/cbioportal/blob/master/core/src/main/java/org/mskcc/cbio/portal/authentication/saml/MSKCCPortalUserDetailsService.java), or copy it and use as a reference.
+    saml.custom.userservice.class=<your_package.your_class_name>
 
-To get MSKCCPortalUserDetailsService.java working with OneLogin, you only need to make two changes:
+:warning: The property `saml.idp.metadata.attribute.email` can also vary per IDP. It is important to set this correctly since this is a required field by the cBioPortal SAML parser (that is, if `org.mskcc.cbio.portal.authentication.saml.SAMLUserDetailsServiceImpl` is chosen for property `saml.custom.userservice.class`). 
 
-1.  Within ```initializeDefaultEmailSuffixes()```, add your email suffixes.  For example:
+:warning: Some IDPs also like to provide their own logout page (e.g. when they don't support the custom SAML Single Logout protocol). For this you can adjust the
```
Comment (pieterlukasse) [inline docs/Authenticating-Users-via-SAML.md]: done

File: docs/Authenticating-Users-via-SAML.md [hunk 6]
```diff
@@ -92,52 +109,68 @@ Within portal.properties, make sure that:
...
+    saml.custom.userservice.class=<your_package.your_class_name>
 
-To get MSKCCPortalUserDetailsService.java working with OneLogin, you only need to make two changes:
+:warning: The property `saml.idp.metadata.attribute.email` can also vary per IDP. It is important to set this correctly since this is a required field by the cBioPortal SAML parser (that is, if `org.mskcc.cbio.portal.authentication.saml.SAMLUserDetailsServiceImpl` is chosen for property `saml.custom.userservice.class`). 
 
-1.  Within ```initializeDefaultEmailSuffixes()```, add your email suffixes.  For example:
+:warning: Some IDPs also like to provide their own logout page (e.g. when they don't support the custom SAML Single Logout protocol). For this you can adjust the 
+`saml.logout.url` to a custom URL provided by the IDP. Also set the `saml.logout.local=true` in this case to indicate that global logout (or Single Logout) is not supported by IDP.
```
Comment (pieterlukasse) [inline docs/Authenticating-Users-via-SAML.md]: done

**Comment by pieterlukasse** [comment]
@n1zea144 thanks for the review! Regarding the changes you requested:
- add `log4j.logger.org.springframework.security=DEBUG`: in my experience this is not necessary if you set the root logger to DEBUG. But it is a good idea to add it as an extra option so one has the option to keep the root logger on INFO and only spring security on DEBUG for example. So I just added it as an extra item and kept the `log4j.category.org.mskcc=DEBUG` option (for the same reason). 
- small text changes: done

Can you please check if all is according to what you want in the last commit? If all is OK, I will squash and let you know.

File: security/security-spring/src/main/java/org/cbioportal/security/spring/authentication/saml/SAMLUserDetailsServiceImpl.java [hunk 7]
```diff
@@ -21,46 +21,52 @@
...
  * @author Pieter Lukasse
  */
+@Service
 public class SAMLUserDetailsServiceImpl implements SAMLUserDetailsService
 {
     private static final Log log = LogFactory.getLog(SAMLUserDetailsServiceImpl.class);
-    private final PortalUserDAO portalUserDAO;
+
+    private static String samlIdpMetadataEmailAttributeName;
+    @Value("${saml.idp.metadata.attribute.email:mail}")
```
Comment (pieterlukasse) [inline security/security-spring/src/main/java/org/cbioportal/security/spring/authentication/saml/SAMLUserDetailsServiceImpl.java]: really nice, I like this

Comment (pieterlukasse) [inline security/security-spring/src/main/java/org/cbioportal/security/spring/authentication/saml/SAMLUserDetailsServiceImpl.java]: :+1:

**Comment by mandawilson** [comment]
Works great and looks great!  Thanks @pieterlukasse!

**Pull request merged**